src/Components/InfiniteScroll.tsx

```diff
@@ -55,37 +69,140 @@ const DisplayPreviousPage = compose<IDisplayPreviousPageProps<any>, IDisplayPrev
+    let previousPlaceHolderHeight = 0;
+    let nextPlaceHolderHeight = 0;
+
+    for (let x = query.response.currentPage - 2; x > 0; x--) {
+      previousPlaceHolderHeight += x in pageHeightMap
+        ? pageHeightMap[x]
+        : estimatedPageHeight;
+    }
+
+    for (let x = query.response.currentPage + 2; x <= query.response.totalPages; x++) {
+      nextPlaceHolderHeight += x in pageHeightMap
+        ? pageHeightMap[x]
+        : estimatedPageHeight;
+    }
+
+    return {
+      previousPlaceHolderHeight,
+      nextPlaceHolderHeight,
+    };
+  }),
+  withProps(({
+    currentPageHeightRef,
+    nextPageHeightRef,
+    pageHeightMap,
+    previousPageHeightRef,
+    recordPageHeight,
+  }) => ({
+    updatePageHeightMap: (query) => {
+      const currentPageHeight = currentPageHeightRef.current.getBoundingClientRect().height;
+      let previousPageHeight;
+      let nextPageHeight;
+
+      const updatedPageHeightMap = {
+        ...pageHeightMap,
+        [query.response.currentPage]: currentPageHeight,
+      };
+
+      if (query.hasPreviousPage()) {
+        previousPageHeight = previousPageHeightRef.current.getBoundingClientRect().height;
+        updatedPageHeightMap[query.response.previousPage] = previousPageHeight;
+      }
+
+      if (query.hasNextPage()) {
+        nextPageHeight = nextPageHeightRef.current.getBoundingClientRect().height;
+        updatedPageHeightMap[query.response.nextPage] = nextPageHeight;
+      }
+
+      recordPageHeight(updatedPageHeightMap);
+      return updatedPageHeightMap;
+    },
+  })),
   withStateHandlers<{ lastScrollTop: number }, { handleScroll }, IInfiniteScrollInternalProps<any>>(
     { lastScrollTop: 0 },
     {
-      handleScroll: ({ lastScrollTop }, { query, updateQuery }) => (clientHeight: number, scrollHeight: number, currentScrollTop: number) => {
-        const scrollingDown = currentScrollTop > lastScrollTop;
-        const scrollBottom = scrollHeight - currentScrollTop - clientHeight;
+      handleScroll: ({ lastScrollTop }, { disableVirtualScrolling, estimatedPageHeight, query, updatePageHeightMap, updateQuery }) =>
+        (clientHeight: number, scrollHeight: number, currentScrollTop: number) => {
+          const scrollingDown = currentScrollTop > lastScrollTop;
+          const currentScrollBottom = scrollHeight - currentScrollTop - clientHeight;
 
-        if (scrollingDown && scrollBottom < (clientHeight / 2) && query.hasNextPage()) {
-          updateQuery(query.getNextPage());
-        }
+          if (!disableVirtualScrolling) {
+            const updatedPageHeightMap = updatePageHeightMap(query);
 
-        return { lastScrollTop: currentScrollTop };
-      },
+            let nextPageToLoad = 1;
+            let nextPageScrollTop = updatedPageHeightMap[1];
+
+            while (nextPageScrollTop < currentScrollTop && nextPageToLoad < query.response.totalPages) {
```

Comment:
Should this be like this instead to make the "seam" in the middle of the visible content instead of the top? It might make scrolling up/down more consistent. i.e. whichever page is "more visible" is the active one.

```
while (nextPageScrollTop < (currentScrollTop + clientHeight/2) && nextPageToLoad < query.response.totalPages)
```

Reply:
good idea! updated